### DISCUSSION TITLE
Real time collaborative beta

### DISCUSSION THREAD

**Discussion by davidjgraph** [jgraph/drawio#2641] [General]
An ongoing problem on both Google Drive and Confluence Cloud has been the latency seeing changes of other users when collaborating. We've avoided calling it "real-time" for this reason. This has been because we used the flle itself to sync changes. OneDrive saves a lot faster than the other two storage options, that's pretty fast already.

We're testing a new architecture that you can test from release 16.6.1 using the fast-sync=1 URL parameter, [https://app.diagrams.net/?fast-sync=1](https://app.diagrams.net/?fast-sync=1). It will only work with Google Drive currently, please don't use it for critical diagrams. We cannot promise there won't be data loss, but it's looking quite stable.

If you get a few minutes to test it out, please add to this thread if you find issues or have improvement suggestions.

**Comment by johanruizb** [comment]
There are problems when dragging and moving the same object almost at the same time. There is a sudden jolt of the object and shifting of the editing view.

Edit: Sorry, I already fixed the link.

[2022-02-17 20-08-26.mp4](https://github.com/johanruizb/file-share/blob/1ce4cd69c951b49b8b59bea265ead4797c3d8b88/2022-02-17%2020-08-26.mp4)

[Direct link](https://github.com/johanruizb/file-share/raw/master/2022-02-17%2020-08-26.mp4)

**Reply by a1812** [reply]
404

**Reply by matbgn** [reply]
Your video is on a private link

**Comment by alderg** [comment]
Sharing cursors:
![share-cursor](https://user-images.githubusercontent.com/921200/154787250-81c5520d-9ed6-43d7-bca6-c4f807ea6f17.gif)

**Reply by mubtakir-lazuardi** [reply]
that's cool bro

**Reply by szpeter80** [reply]
Holy moly. This is awesome!

**Reply by AmauryOrtega** [reply]
Looks amazing

**Reply by spencerframe** [reply]
This is absolutely incredible! A real-time interactive diagram would save me hours on my projects! It worked the first time, but after trying to get to another tab in the file, ran into an error. Upon refresh, we can't see cursors or real-time anymore. Got a few bugs to work through, but you guys nailed it on the head. Love it!

**Reply by davidjgraph** [reply]
@suspence21 maybe you weren't on 16.6.1 after a refresh. Might be worth clearing out the service worker. https://app.diagrams.net/clear.html

**Reply by spencerframe** [reply]
@davidjgraph My co-worker and I both tried clearing the service worker, refresh, hard-refresh. We double-checked to make sure we were both on 16.6.2, and had the fast-sync parameter. We'll give it another try next week after my vacation.

**Comment by spencerframe** [comment]
How do I update to 16.6.1 in Google Drive. I've checked everywhere. I'm stuck on 16.5.6

**Reply by alderg** [reply]
Close all tabs and visit https://app.diagrams.net/clear.html

**Reply by alderg** [reply]
@suspence21 Is it working now? If not, are there any errors in the browser console?

**Reply by spencerframe** [reply]
Yes, it worked!

**Comment by oscarleonapp** [comment]
Muchas gracias por el trabajo realizado, esta genial la opción en tiempo real.

**Comment by alderg** [comment]
Now available for OneDrive in v16.6.3 https://app.diagrams.net/?mode=onedrive
Close tabs and visit https://app.diagrams.net/clear.html to update.

**Comment by charlie-amihan** [comment]
Appending ?fast-sync=1 (or &fast-sync=1) to a shared diagram I can normally load switches to 'Error loading file
The file was not found. It does not exist or you do not have access.'

Version: 16.6.3

(Note: there was some weirdness where my original attempt to load the original link threw the same error but after a couple minutes without further permission changes I was able to access it, unsure if related).

**Reply by alderg** [reply]
What exact URL did you end up with after appending ?fast-sync, ie was the #G... before or after fast-sync (it must be before, ie. app.diagrams.net/?fast-sync=1#G...).

**Reply by charlie-amihan** [reply]
Oh hey that is indeed how URLs work, missed the hash ^^. Thanks

**Comment by alderg** [comment]
Next release will allow to enable real time collaboration per-file in Google Drive (via File, Properties):
![collaboration-switch](https://user-images.githubusercontent.com/921200/155531377-7486925a-5234-4b24-8185-22711de77439.gif)

**Comment by zibik70** [comment]
After saving a file to OneDrive, the data permanently disappears on reopening, I've already weighed the same diagram 3 times, very annoying.

**Reply by alderg** [reply]
Use fast-sync=0 to disable fast sync for OneDrive while we're working on a fix.

**Reply by zibik70** [reply]
Where I can find this switch?

**Reply by zibik70** [reply]
https://app.diagrams.net/?fast-sync=0 is well?

**Reply by m-mohamedin** [reply]
yes

**Reply by alderg** [reply]
Did you use multiple clients concurrently when this happened? If yes, did they edit concurrently? What browser / device did you use? Did you edit using one more more iPads? If you could answer these questions, that will help us narrow down the problem. Meanwhile we'll disable fast sync in OneDrive by default (can be enable with fast-sync=1) in the next release until the problem is resolved.

**Reply by zibik70** [reply]
No, It was only my open diagrams, I use Firefox 97.0.1.

**Reply by alderg** [reply]
And you only used one browser? Did you get any error messages? Are you using Windows 11? Did the shapes disappear while using the app or only after you reloaded the file?

**Reply by zibik70** [reply]
Yes, I opened file from explorer (files was in OneDrive) and from firefox, in every case once opened draw was exipre and file had 1kb. I'm using windows 10 Pro 21H2. The shapes disappers after reloaded the file.

**Reply by davidjgraph** [reply]
That is a different issue. OneDrive has a file history, you can restore older versions from the OneDrive page.

**Reply by alderg** [reply]
So you opened the file from the local file via "Device"? Did it happen with all files? Does it only happen if you use fast-sync=1 in the URL? When you reopened the file, how did you reopen it?

**Comment by gaos17** [comment]
My files were saved in Chrome browser before, but NOW I can't find them. I wonder if this has something to do with it?  Can I get my files back?

**Reply by davidjgraph** [reply]
No, that is unrelated.

**Reply by gaos17** [reply]
> 

thanks, Can i get my files? I only saved in Chrome browser.

**Reply by gaos17** [reply]
> 
I remember. I deleted my browser data a week ago.  I don't think the file will ever be recovered

**Comment by camherry** [comment]
Hello,
We use the draw.io application from Nextcloud, is it possible to activate this feature on this platform?
Thanks in advance for your feedback,
Best regards,

**Reply by davidjgraph** [reply]
No, we're not the authors of the nextcloud app, you would need to talk to them.

**Reply by camherry** [reply]
Thank you for your reply.
So it would be an adaptation of the Nextcloud platform for it to work?

**Reply by davidjgraph** [reply]
Sorry, no idea, I've never developed on nextcloud.

**Reply by camherry** [reply]
I saw on another Github ticket that you are writing a replacement that isn't dependent on any particular system. Currently, it needs if-match/etag header support on the file to work.
Would this be the solution ?
https://github.com/pawelrojek/nextcloud-drawio/issues/14#issuecomment-441993299

**Comment by mawid6** [comment]
Hi! Is it in your roadmap to have a similar realtime solution for confluence?

**Reply by davidjgraph** [reply]
Confluence cloud, yes. On-prem we're blocked by https://community.developer.atlassian.com/t/real-time-communication-for-confluence-dc-plugin-websocket-sse/56322 . If you're on-prem, never hurts to tell Atlassian this affects you.

**Reply by mawid6** [reply]
Thanks, sounds good! We're on cloud, so that's good for me 👍

**Comment by davidjgraph** [comment]
This is the default on Google Drive and OneDrive for a few weeks now. If you're not seeing it, you probably need to update to the latest version, https://app.diagrams.net/clear.html

**Comment by mapl** [comment]
Are other collaboration implementations planned? For instance, using a Windows SMB File share?

**Reply by davidjgraph** [reply]
Using the desktop app? I think that should work reasonably well, just no remote cursors and selection.

**Comment by dstephan1** [comment]
There are issues in Google Drive with files appearing to save but not actually save.

See [code](https://github.com/jgraph/drawio/blob/236e38393185a86ceea337bb17b4eaa7b2dd37e2/src/main/webapp/js/diagramly/DriveFile.js#L603) 

If files don't have the 'collaboration' property set, this returns _undefined_, and isRealtimeEnabled() returns true, even if it shouldn't be. This is happening for me because there is no RT server running. I'm guessing that if the app.diagrams.net/rt goes down, others might run into the same issue.

**Reply by davidjgraph** [reply]
OK thanks, we'll take a look. Do you not run the collab server because it's too much effort to run the VM? We run it on Google App Engine, you can use that if easier.

**Comment by davidjgraph** [comment]
Confluence Cloud switched to new system in 17.3.0.

**Comment by mark-james** [comment]
Is it possible to use the new collaborative features in embed mode?

**Reply by alderg** [reply]
No, that's currently not supported.

**Reply by mark-james** [reply]
Thanks, Gaudenz. Is this something on the roadmap or not a use case you plan to pursue?

**Reply by alderg** [reply]
Not on the roadmap and there are some requirements for the backend for this to work (ie. if-match header). Please open a feature request, this helps us to check if there is demand for this feature based on votes.
